Tools/clinic/clinic.py

```diff
@@ -2727,18 +2725,38 @@ class object_converter(CConverter):
     format_unit = 'O'
 
     def converter_init(self, *, converter=None, type=None, subclass_of=None):
+        self.subclass_of = subclass_of
         if converter:
             if subclass_of:
                 fail("object: Cannot pass in both 'converter' and 'subclass_of'")
             self.format_unit = 'O&'
             self.converter = converter
-        elif subclass_of:
-            self.format_unit = 'O!'
-            self.subclass_of = subclass_of
 
         if type is not None:
             self.type = type
 
+    def modify(self):
+        if not self.subclass_of:
+            return ""
+        err_check = '!PyType_IsSubtype(Py_TYPE({}), {})'.format(self.name, self.subclass_of)
+
+        # Shortcut subtype checks for types that can use PyType_FastSubclass
+        if self.subclass_of == '&PyTuple_Type':
+            err_check = '!PyTuple_Check({})'.format(self.name)
+        elif self.subclass_of == '&PyDict_Type':
+            err_check = '!PyDict_Check({})'.format(self.name)
```

Comment:
Why only tuple and dict? Same fast path would apply to all types with fast subclass flags from object.h ( https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Include/object.h#L660 ), including int, list, bytes, str, BaseException and type.